architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+
+	// Scope is the resource ID of the scope.
+	// +kubebuilder:validation:Optional
+	Scope string `json:"scope,omitempty"`
+
+	// Resource is the resource ID of the deployment.
+	// +kubebuilder:validation:Optional
+	Resource string `json:"resource,omitempty"`
+
+	// Operation tracks the status of an in-progress provisioning operation.
+	// +kubebuilder:validation:Optional
+	Operation *ResourceOperation `json:"operation,omitempty"`
+
+	// Phrase indicates the current status of the Bicep Deployment.
+	// +kubebuilder:validation:Optional
+	Phrase BicepDeploymentPhrase `json:"phrase,omitempty"`
+
+  // Description is a human-readable description of the status of the Bicep Deployment.
+  // +kubebuilder:validation:Optional
+  Description string `json:"description,omitempty"`
+}
+```
+
+#### `rad bicep generate bicepdeployment` Command
+
+The `rad bicep generate bicepdeployment <bicep file>` command will be a new command that will generate an BicepDeployment resource when provided a Bicep template and (optionally) a Bicep parameters file. The command will take the following arguments:
+
+- `-p, --parameters`: The path to the Bicep parameters file
+- `-o, --output`: The path to the output file where the BicepDeployment resource will be written
+
+The command will read the Bicep template and parameters files and generate a BicepDeployment resource with the template and parameters set. The command will write the BicepDeployment resource to the output file.
+
+#### BicepDeployment Controller
+
+The `radius-controller` will be updated to include a new controller that reconciles BicepDeployment resources. When an BicepDeployment resource is created or updated, the controller will compile the Bicep template and parameters into ARM JSON and send the ARM JSON to the Radius API to deploy the resources. The controller will update the `status.phrase` and the `status.description` of the BicepDeployment resource based on the status of the deployment (see `BicepDeploymentPhrase` above).
+
+The controller will also perform some simple lookups to Radius to determine if the provided template can be deployed. It will make two checks:
+- Check if the template contains an empty `param environment`. If it does, the controller will query the Radius API to get an environment ID (must have exactly one environment if left unspecified like this) and deploy the template using that environment.
+- Check if the template contains a `param application`. If it does, the controller will query the Radius API to get an application ID (must have exactly one application if left unspecified like this) and deploy the template using that application.
+
+The controller will also be shipped with a Bicep binary which will allow it to compile Bicep templates and parameters into ARM JSON. The Bicep binary will be included in the `radius-controller` container image.
+
+### Alternatives Considered
+
+- We could create a separate service for Bicep compilation.
```

Comment:
We could also do compilation as part of the CLI command, or as part of the GitOps integration (Flux Controller + ArgoCD custom tool).